File: CryptoApp/Shared Components/SkeletonView/AddCoin/AddCoinCellSkeleton.m
//
//  AddCoinCellSkeleton.m
//  CryptoApp
//
//  Created by Jansen Castillo
//

#import "AddCoinCellSkeleton.h"
#import "SkeletonView.h"

@interface AddCoinCellSkeleton ()

@property (nonatomic, strong) NSArray<SkeletonView *> *skeletonViews;

// Layout components to match AddCoinCell
@property (nonatomic, strong) SkeletonView *imageSkeleton;
@property (nonatomic, strong) SkeletonView *symbolSkeleton;
@property (nonatomic, strong) SkeletonView *nameSkeleton;

@end

@implementation AddCoinCellSkeleton

// MARK: - Initialization

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self setupUI];
        [self collectSkeletonViews];
    }
    return self;
}

- (instancetype)initWithCoder:(NSCoder *)coder {
    self = [super initWithCoder:coder];
    if (self) {
        [self setupUI];
        [self collectSkeletonViews];
    }
    return self;
}

// MARK: - Setup

- (void)setupUI {
    [self createSkeletonViews];
    [self configureCellAppearance];
    [self layoutSkeletonViews];
    [self applyConstraints];
}

- (void)createSkeletonViews {
    self.imageSkeleton = [SkeletonView circleSkeletonWithDiameter:40];
    self.symbolSkeleton = [SkeletonView textSkeletonWithWidth:60 height:16];
    self.nameSkeleton = [SkeletonView textSkeletonWithWidth:80 height:14];
}

- (void)configureCellAppearance {
    // Match AddCoinCell styling
    self.backgroundColor = [UIColor systemBackgroundColor];
    self.layer.cornerRadius = 8.0;
    self.layer.borderWidth = 1.0;
    self.layer.borderColor = [UIColor colorWithRed:0.9 green:0.9 blue:0.9 alpha:1.0].CGColor;
}

- (void)layoutSkeletonViews {
    [self.contentView addSubview:self.imageSkeleton];
    [self.contentView addSubview:self.symbolSkeleton];
    [self.contentView addSubview:self.nameSkeleton];
}

- (void)applyConstraints {
    // Layout to match AddCoinCell (horizontal layout with image on left, text on right)
    [NSLayoutConstraint activateConstraints:@[
        // Image skeleton positioning (left side, centered vertically)
        [self.imageSkeleton.leadingAnchor constraintEqualToAnchor:self.contentView.leadingAnchor constant:16],
        [self.imageSkeleton.centerYAnchor constraintEqualToAnchor:self.contentView.centerYAnchor],
        
        // Symbol skeleton positioning (right of image, top)
        [self.symbolSkeleton.leadingAnchor constraintEqualToAnchor:self.imageSkeleton.trailingAnchor constant:12],
        [self.symbolSkeleton.topAnchor constraintEqualToAnchor:self.contentView.topAnchor constant:12],
        [self.symbolSkeleton.trailingAnchor constraintLessThanOrEqualToAnchor:self.contentView.trailingAnchor constant:-50], // Leave space for potential checkmark
        
        // Name skeleton positioning (right of image, below symbol)
        [self.nameSkeleton.leadingAnchor constraintEqualToAnchor:self.imageSkeleton.trailingAnchor constant:12],
        [self.nameSkeleton.topAnchor constraintEqualToAnchor:self.symbolSkeleton.bottomAnchor constant:2],
        [self.nameSkeleton.trailingAnchor constraintLessThanOrEqualToAnchor:self.contentView.trailingAnchor constant:-50],
        [self.nameSkeleton.bottomAnchor constraintLessThanOrEqualToAnchor:self.contentView.bottomAnchor constant:-12]
    ]];
}

- (void)collectSkeletonViews {
    self.skeletonViews = @[
        self.imageSkeleton,
        self.symbolSkeleton,
        self.nameSkeleton
    ];
}

// MARK: - Public Methods

- (void)startShimmering {
    for (SkeletonView *skeletonView in self.skeletonViews) {
        [skeletonView startShimmering];
    }
}

- (void)stopShimmering {
    for (SkeletonView *skeletonView in self.skeletonViews) {
        [skeletonView stopShimmering];
    }
}

// MARK: - Reuse

- (void)prepareForReuse {
    [super prepareForReuse];
    [self stopShimmering];
}

+ (NSString *)reuseID {
    return @"AddCoinCellSkeleton";
}

@end

File: CryptoApp/Shared Components/SkeletonView/AddCoinCellSkeleton.swift
import UIKit

final class AddCoinCellSkeleton: UICollectionViewCell {
    
    // MARK: - Properties
    
    private var skeletonViews: [SkeletonView] = []
    
    // Layout components to match AddCoinCell
    private let imageSkeleton = SkeletonView.circleSkeleton(diameter: 40)
    private let symbolSkeleton = SkeletonView.textSkeleton(width: 60, height: 16)
    private let nameSkeleton = SkeletonView.textSkeleton(width: 80, height: 14)
    
    // MARK: - Initialization
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupUI()
        collectSkeletonViews()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        setupUI()
        collectSkeletonViews()
    }
    
    // MARK: - Setup
    
    private func setupUI() {
        // Match AddCoinCell styling
        backgroundColor = UIColor.systemBackground
        layer.cornerRadius = 8
        layer.borderWidth = 1
        layer.borderColor = UIColor(red: 0.9, green: 0.9, blue: 0.9, alpha: 1.0).cgColor
        
        contentView.addSubview(imageSkeleton)
        contentView.addSubview(symbolSkeleton)
        contentView.addSubview(nameSkeleton)
        
        // Layout to match AddCoinCell (horizontal layout with image on left, text on right)
        NSLayoutConstraint.activate([
            // Image skeleton positioning (left side, centered vertically)
            imageSkeleton.leadingAnchor.constraint(equalTo: contentView.leadingAnchor, constant: 16),
            imageSkeleton.centerYAnchor.constraint(equalTo: contentView.centerYAnchor),
            
            // Symbol skeleton positioning (right of image, top)
            symbolSkeleton.leadingAnchor.constraint(equalTo: imageSkeleton.trailingAnchor, constant: 12),
            symbolSkeleton.topAnchor.constraint(equalTo: contentView.topAnchor, constant: 12),
            symbolSkeleton.trailingAnchor.constraint(lessThanOrEqualTo: contentView.trailingAnchor, constant: -50), // Leave space for potential checkmark
            
            // Name skeleton positioning (right of image, below symbol)
            nameSkeleton.leadingAnchor.constraint(equalTo: imageSkeleton.trailingAnchor, constant: 12),
            nameSkeleton.topAnchor.constraint(equalTo: symbolSkeleton.bottomAnchor, constant: 2),
            nameSkeleton.trailingAnchor.constraint(lessThanOrEqualTo: contentView.trailingAnchor, constant: -50),
            nameSkeleton.bottomAnchor.constraint(lessThanOrEqualTo: contentView.bottomAnchor, constant: -12)
        ])
    }
    
    private func collectSkeletonViews() {
        skeletonViews = [
            imageSkeleton,
            symbolSkeleton,
            nameSkeleton
        ]
    }
    
    // MARK: - Public Methods
    
    func startShimmering() {
        skeletonViews.forEach { $0.startShimmering() }
    }
    
    func stopShimmering() {
        skeletonViews.forEach { $0.stopShimmering() }
    }
    
    // MARK: - Reuse
    
    override func prepareForReuse() {
        super.prepareForReuse()
        stopShimmering()
    }
    
    static var reuseID: String {
        return "AddCoinCellSkeleton"
    }
} 